[Bug] Unit test cases based on the pg_container fixture fail under parallel pytest execution

**Bug Description**
TestSandboxMetaStoreWithDocker & TestSandboxTableWithPostgres failed under parallel pytest execution.

**Steps to Reproduce**
run pytest 


**Expected Behavior**
All cases should pass.

**Actual Behavior**
Some cases fail.

**Error Logs**
```bash
uv run pytest -n 10 tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres

...... (omitted)

[gw9] ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_get_nonexistent_returns_none 
[gw8] ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_json_fields_postgresql 
[gw5] ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_insert_duplicate_raises 2026-04-14 02:37:28,806 DEBUG selector_events.py:54 -- Using selector: EpollSelector
2026-04-14 02:37:28,806 DEBUG selector_events.py:54 -- Using selector: EpollSelector
2026-04-14 02:37:28,813 DEBUG selector_events.py:54 -- Using selector: EpollSelector
2026-04-14 02:37:28,814 DEBUG selector_events.py:54 -- Using selector: EpollSelector

[gw3] ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_fixture_connection 2026-04-14 02:37:28,857 DEBUG selector_events.py:54 -- Using selector: EpollSelector
2026-04-14 02:37:28,858 DEBUG selector_events.py:54 -- Using selector: EpollSelector

[gw1] ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_delete 2026-04-14 02:37:28,945 DEBUG selector_events.py:54 -- Using selector: EpollSelector
2026-04-14 02:37:28,946 DEBUG selector_events.py:54 -- Using selector: EpollSelector

[gw4] PASSED tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_update 
[gw7] PASSED tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_list_by_rejects_blacklisted_column 
[gw6] PASSED tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_insert_and_get 
[gw0] PASSED tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_list_by 

============================================================================ ERRORS =============================================================================
_______________________________________ ERROR at setup of TestSandboxTableWithPostgres.test_get_nonexistent_returns_none ________________________________________
[gw9] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/admin/core/test_sandbox_table.py:111: in db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
E   ConnectionError: unexpected connection_lost() call
__________________________________________ ERROR at setup of TestSandboxTableWithPostgres.test_json_fields_postgresql ___________________________________________
[gw8] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/admin/core/test_sandbox_table.py:111: in db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
E   ConnectionError: unexpected connection_lost() call
__________________________________________ ERROR at setup of TestSandboxTableWithPostgres.test_insert_duplicate_raises __________________________________________
[gw5] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/admin/core/test_sandbox_table.py:111: in db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
E   ConnectionError: unexpected connection_lost() call
____________________________________________ ERROR at setup of TestSandboxTableWithPostgres.test_fixture_connection _____________________________________________
[gw3] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/admin/core/test_sandbox_table.py:111: in db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/asyncio/selector_events.py:995: in _read_ready__data_received
    data = self._sock.recv(self.max_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   ConnectionResetError: [Errno 104] Connection reset by peer
__________________________________________________ ERROR at setup of TestSandboxTableWithPostgres.test_delete ___________________________________________________
[gw1] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/admin/core/test_sandbox_table.py:111: in db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/asyncio/selector_events.py:995: in _read_ready__data_received
    data = self._sock.recv(self.max_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   ConnectionResetError: [Errno 104] Connection reset by peer
==================================================================== short test summary info ====================================================================
ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_get_nonexistent_returns_none - ConnectionError: unexpected connection_lost() call
ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_json_fields_postgresql - ConnectionError: unexpected connection_lost() call
ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_insert_duplicate_raises - ConnectionError: unexpected connection_lost() call
ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_fixture_connection - ConnectionResetError: [Errno 104] Connection reset by peer
ERROR tests/unit/admin/core/test_sandbox_table.py::TestSandboxTableWithPostgres::test_delete - ConnectionResetError: [Errno 104] Connection reset by peer
================================================================= 4 passed, 5 errors in 16.59s ==================================================================
```


```
uv run pytest -n 10 tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker

...... (omitted)

[gw1] PASSED tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_list_by_queries_db 
[gw0] ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_save_writes_redis_and_db 
[gw5] ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_iter_alive_sandbox_ids 
[gw2] ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_remove_deletes_redis_and_db 
[gw3] ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_update_writes_redis_and_db 

============================================================================ ERRORS =============================================================================
________________________________________ ERROR at setup of TestSandboxMetaStoreWithDocker.test_save_writes_redis_and_db _________________________________________
[gw0] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/sandbox/test_sandbox_meta_store.py:388: in real_db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
E   ConnectionError: unexpected connection_lost() call
_________________________________________ ERROR at setup of TestSandboxMetaStoreWithDocker.test_iter_alive_sandbox_ids __________________________________________
[gw5] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/sandbox/test_sandbox_meta_store.py:388: in real_db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
E   ConnectionError: unexpected connection_lost() call
_______________________________________ ERROR at setup of TestSandboxMetaStoreWithDocker.test_remove_deletes_redis_and_db _______________________________________
[gw2] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/sandbox/test_sandbox_meta_store.py:388: in real_db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/asyncio/selector_events.py:995: in _read_ready__data_received
    data = self._sock.recv(self.max_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   ConnectionResetError: [Errno 104] Connection reset by peer
_______________________________________ ERROR at setup of TestSandboxMetaStoreWithDocker.test_update_writes_redis_and_db ________________________________________
[gw3] linux -- Python 3.11.4 /dev_docker_dir/rock_features/db/.venv/bin/python3
tests/unit/sandbox/test_sandbox_meta_store.py:388: in real_db
    await provider.create_tables()
rock/admin/core/db_provider.py:40: in create_tables
    async with self.engine.begin() as conn:
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/contextlib.py:204: in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:1068: in begin
    async with conn:
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/base.py:121: in __aenter__
    return await self.start(is_ctxmanager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py:275: in start
    await greenlet_spawn(self.sync_engine.connect)
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:201: in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3277: in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:143: in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py:3301: in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:447: in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:1264: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:711: in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:177: in _do_get
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/impl.py:175: in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:388: in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:673: in __init__
    self.__connect()
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:899: in __connect
    with util.safe_reraise():
.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:224: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py:895: in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py:661: in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py:629: in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:955: in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:132: in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:196: in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connection.py:2443: in connect
    return await connect_utils._connect(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1249: in _connect
    raise last_error or exceptions.TargetServerAttributeNotMatched(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1218: in _connect
    conn = await _connect_addr(
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1054: in _connect_addr
    return await __connect_addr(params, True, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:1099: in __connect_addr
    tr, pr = await connector
             ^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/asyncpg/connect_utils.py:977: in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
                     ^^^^^^^^^^^^^^^^
/root/.local/share/uv/python/cpython-3.11-linux-x86_64-gnu/lib/python3.11/asyncio/selector_events.py:995: in _read_ready__data_received
    data = self._sock.recv(self.max_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   ConnectionResetError: [Errno 104] Connection reset by peer
==================================================================== short test summary info ====================================================================
ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_save_writes_redis_and_db - ConnectionError: unexpected connection_lost() call
ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_iter_alive_sandbox_ids - ConnectionError: unexpected connection_lost() call
ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_remove_deletes_redis_and_db - ConnectionResetError: [Errno 104] Connection reset by peer
ERROR tests/unit/sandbox/test_sandbox_meta_store.py::TestSandboxMetaStoreWithDocker::test_update_writes_redis_and_db - ConnectionResetError: [Errno 104] Connection reset by peer
================================================================= 1 passed, 4 errors in 14.57s ==================================================================
```

**Component Affected**
- [ ] Sandbox
- [ ] Actions
- [ ] Deployments
- [ ] SDK & API
- [ ] Envhub
- [ ] CLI 
- [ ] Performance & Optimization
- [ ] Documentation & Examples
- [x] Testing

--- a/tests/unit/conftest.py
+++ b/tests/unit/conftest.py
@@ -353,6 +353,15 @@ def pg_container():
         else:
             raise TimeoutError("PostgreSQL container did not become ready within 30s")
 
+        # Confirm with a real SQL query to close the startup race window.
+        while _t.time() < deadline:
+            code, _ = container.exec_run(f'psql -U {_PG_USER} -d {_PG_DB} -c "SELECT 1"')
+            if code == 0:
+                break
+            _t.sleep(0.5)
+        else:
+            raise TimeoutError("PostgreSQL: pg_isready OK but SELECT 1 still failing")
+
         host, port = _docker_resolve_host_port(container, network_name, _PG_PORT)
         yield {
             "host": host, "port": port,

